kernel/src/process_standard.rs

```diff
@@ -614,7 +616,16 @@ impl<C: Chip> Process for ProcessStandard<'_, C> {
                 let old_break = self.app_break.get();
                 self.app_break.set(new_break);
                 self.chip.mpu().configure_mpu(config);
-                Ok(old_break)
+
+                let base = self.mem_start() as usize;
+                let break_result = CapabilityPtr::new_with_metadata(
+                    old_break as *const (),
+                    base,
+                    (new_break as usize) - base,
+                    MetaPermissions::ReadWrite,
+                );
```

Comment:
OK, so on legacy architectures this diff is basically a noop (the last 3 arguments are literally ignored and the implementation is just a type case) and we'd expect to compile away. On a future CHERI this will matter. Semantically it's saying this new break is the _end_ of an accesible read/write region that starts at memory base.

Reply:
I presume that on a CHERI platform, this code is a crucial part of the enforcement of memory isolation in the system, in lieu of having an MPU that enforces memory isolation. Is that correct?

If it is, I think this should be an unsafe operation, in the presence of APIs that take `CapabilityPtr`s and provide them to userspace.

Reply:
To clarify @lschuermann's suggestion, I think he means to propose that `new_with_metadata` be marked unsafe, which I think is reasonable. For example, capsules, which are not allowed to use unsafe, should not generally be able to fabric a valid userspace pointer at out of thin air (this will be particularly important when CHERI implementations use this as the sole means of sandboxing userspace memory accesses.

Reply:
I think this is a very important  philosophical point which warrants some discussion. Do we mark all interfaces that influence isolation as unsafe? 

Arguably, an equivalent method  on non-CHERI (as it might grant the user authority at a hardware level, but does not create any new references in the type system) is `allocate_region` on the MPU trait. It is very "unsafe" in the colloquial sense, but is not marked as such, and is arguably not unsafe in the Rust sense. We do not break memory safety by calling this method (although the user might!). But really, the unsafe behaviour is there in userspace.

Given `new_with_metadata` does _not_ implement Deref, from Rust's point of view,  its construction is safe.

If we did decide that all isolation enforcing code should have unsafe interfaces (because we are relying on it for our "context switch" to be sound), I think the boundary might extend further than one might expect and is not the remit of this PR.

For now, I don't think this one should be "special".

Reply:
I'm trying my best to not open yet another can of worms here. The short version is that we also rely on crate dependencies and visibility for safety (where we should is another question). Because this is a public type and semi-trusted code (i.e., capsules) can name it, it should be unsafe. Safe capsules *must not* be able to endanger isolation.

I don't think that the MPU interface is a great inspiration here; it's well possible that it has similar problems which we should revisit at another point.

Reply:
Turned this into an unsafe method, as per the reasoning above.